### DISCUSSION TITLE
transfer: upload performance; avoid tiny send - Side effect

### DISCUSSION THREAD

**Discussion by baukus** [curl/curl#10217] [Q&A]
This change https://github.com/curl/curl/commit/7f43f3dc5994d01b12eb1dfe2ebf5e8371b4e32b has a side effect that breaks a client upload application that uploads files/objects in chunks as dictated by the server. A 3mb file is uploaded in 3 1mb chunks; each chunk is a separate call to  curl_easy_perform().

For the purposes of _readwrite_upload(),_ the chunk size is expressed as  _data->state.infilesize_.

When this change kicks in because _k->upload_present < curl_upload_refill_watermark(data)_ then everything appears to operate correctly until, many sends later,  the upload approaches the 1mb chunk limit. 

Suppose _offset = k->upload_present = 1534_ then many sends later the upload at 1mb - 1534 works as expected but the following upload is not 1534, but rather a full 64k and curl continues to send data well beyond the 1mb limit. The server stops reading at 1mb, the tcp window closes, the socket backs up, and finally the client closes the connection via a timeout. The upload accounting is always off by the original _offset = k->upload_present_.

What I cannot figure out is where/how this well contained change leaks out and causes the upload accounting to be off by _offset_:

```
        k->upload_fromhere += offset;
        result = Curl_fillreadbuffer(data, data->set.upload_buffer_size-offset,
                                     &fillcount);
        k->upload_fromhere -= offset;
```
Any help tracking this down is appreciated.

**Comment by bagder** [comment]
> a side effect that breaks a client upload application that uploads files/objects in chunks as dictated by the server.

(If you think this is an issue, then file an issue.)

I don't think libcurl guarantees any chunk size to be maintained or get sent as-is so if it worked before that was more good luck than anything else. Also, proxies and networks might make it end up differently in the other end anyway.
